server/service/src/repack/validate.rs

```diff
@@ -0,0 +1,65 @@
+use repository::{
+    EqualFilter, RepositoryError, StockLine, StockLineFilter, StockLineRepository, StockLineRow,
+    StorageConnection,
+};
+
+use super::{
+    common::calculate_stock_line_total,
+    insert::{InsertRepack, InsertRepackError},
+};
+
+pub fn validate(
+    connection: &StorageConnection,
+    store_id: &str,
+    input: &InsertRepack,
+) -> Result<StockLineRow, InsertRepackError> {
+    let stock_line = match check_stock_line_exists(connection, store_id, &input.stock_line_id)? {
+        Some(stock_line) => stock_line,
+        None => return Err(InsertRepackError::StockLineDoesNotExist),
+    };
+
+    if store_id != stock_line.stock_line_row.store_id {
+        return Err(InsertRepackError::NotThisStoreStockLine);
+    };
+
+    if check_packs_are_fractional(input) {
+        return Err(InsertRepackError::CannotHaveFractionalRepack);
+    }
+
+    if check_stock_line_reduced_to_zero(input, &stock_line.stock_line_row) {
+        return Err(InsertRepackError::StockLineReducedBelowZero(stock_line));
+    }
+
+    Ok(stock_line.stock_line_row)
+}
+
+fn check_stock_line_exists(
+    connection: &StorageConnection,
+    store_id: &str,
+    id: &str,
+) -> Result<Option<StockLine>, RepositoryError> {
+    Ok(StockLineRepository::new(connection)
+        .query_by_filter(
+            StockLineFilter::new().id(EqualFilter::equal_to(id)),
+            Some(store_id.to_string()),
+        )?
+        .pop())
+}
+
+fn check_stock_line_reduced_to_zero(input: &InsertRepack, stock_line: &StockLineRow) -> bool {
+    let stock_line_total = calculate_stock_line_total(stock_line);
+
+    if stock_line_total < input.number_of_packs {
```

Comment:
hmm, i might have been confused in my previous comment, is number_of_packs in input, number_of_packs of the original stock line ? if so this should just be comparing against that rather then total quantity

Reply:
I'm confused by your comment 😆 
I've refactored this anyways~